File: Implementations/UnionFind.cpp
// Union-Find Implementation

#include <iostream>
#include <algorithm>
#include <vector>
#define MAXN 10000
 
using namespace std;

int N, M;
int arr[MAXN], sz[MAXN];

void init() {
	for(int i = 0; i < N; i++) {
		arr[i] = i;
		sz[i] = 1;
	}
}

int root(int a) {
	while(arr[a] != a) {
		arr[a] = arr[arr[a]];
		a = arr[a];
	}
	return a;
}

void merge(int a, int b) {
	int r_a = root(a);
	int r_b = root(b); 
	if(r_a == r_b) return;
	
	if(sz[r_a] < sz[r_b]) {
		arr[r_a] = r_b;
		sz[r_b] += sz[r_a];
	} else {
		arr[r_b] = r_a;
		sz[r_a] += sz[r_b];
	}
}

int main() {
	cin >> N >> M;
	init();

	for(int i = 0; i < M; i++) {
		int a, b;
		a--; b--;
		cin >> a >> b;
		merge(a, b);
	}
}

File: Implementations/UnionFind.java
// Union-Find Implementation

import java.util.*;
import java.io.*;

public class UnionFind {
	static int N;
	static int[] rt, sz;
	public static void main(String[] args) throws IOException {
		String file = "";
		// BufferedReader br = new BufferedReader(new FileReader(new File(file + ".in")));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		rt = new int[N];
		sz = new int[N];
		init();
		
		int M = Integer.parseInt(st.nextToken());
		for(int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			merge(Integer.parseInt(st.nextToken())-1, Integer.parseInt(st.nextToken())-1);
		}
		
		br.close();
	}
	
	public static void init() {
		for(int i = 0; i < N; i++) {
			rt[i] = i;
			sz[i] = 1;
		}
	}
	
	public static int root(int a) {
		while(rt[a] != a) {
			rt[a] = rt[rt[a]];
			a = rt[a];
		}
		
		return a;
	}
	
	public static void merge(int a, int b) {
		int r_a = root(a), r_b = root(b);
		if(r_a == r_b) return;
		
		if(sz[r_a] < sz[r_b]) {
			rt[r_a] = r_b;
			sz[r_b] += sz[r_a];
		} else {
			rt[r_b] = r_a;
			sz[r_a] += sz[r_b];
		}
	}
}
